numba_cuda/numba/cuda/dispatcher.py

```diff
@@ -115,6 +118,10 @@ def __init__(self, py_func, argtypes, link=None, debug=False,
                                              'cpp_function_wrappers.cu')
             link.append(functions_cu_path)
 
+        if "NRT_" in asm:
```

Comment:
This code is a quick and dirty way to identify when the nrt c++ functions need to be added to the link, but there's room for improvement here. 

High level, I usually try and avoid having to inspect something that I created, because it seems better to capture whatever the needed information is during the creation process. So in this case my instinct leans towards deciding to link at the time the NRT functions are inserted rather than here. 

However numba has the additional optimization pass that can prune out NRT API calls, meaning that we could end up in a situation where we decide to link because NRT calls were inserted, but ultimately don't need them because they were optimized out.

This pushes me back towards inspecting the final object, though I think we should inspect the most optimized possible state of the asm in the end. An improvement beyond what's written might be to check if the full function names are contained in the asm, but I still think it's possible to break that. 

@gmarkall could we inspect some of the intermediate files that come after the PTX stage, but before link time, maybe for certain symbols, and decide to link or not based on that?

Reply:
why not respect `enable_nrt = True` in `CUDATargetContext` to decide if link `nrt.cu`?

I think if the original Numba IR has `NRT_..` calls, the optimized version won't remove them all in most cases.

Reply:
A good point fron @dlee992 - If we link the NRT functions when NRT is enabled and they turn out not to be needed, then they'll be dropped by the linker. So maybe it is just fine to just link the NRT code all the time if we have NRT enabled.

Reply:
I'm thinking how to port https://github.com/numba/numba/pull/8458 to numba-cuda. 

I think more `xxx.cu` will show up, perhaps you can revise the linking process to a generic way, as Testhound did with `link_to_library_functions`. Does it make  sense?
```python
        # Link to the helper library functions if needed
        link_to_library_functions(helper_functions, 'cuda_helperlib.cu')
        # Link to the CUDA FP16 math library functions if needed
        res = link_to_library_functions(cuda_fp16_math_funcs,
                                        'cpp_function_wrappers.cu',
                                        '__numba_wrapper_')
```

cc @gmarkall

Reply:
I think that would be OK for now... Thinking about the trajectory where more and more .cu files show up, we should maybe have another mechanism adding files to the link as we generate the calls to them, but for now I think using the `link_to_library_functions` method that scans the assembly will be OK and allows us to move forward.